### PR TITLE
Replace base abstract classes in API modules with interfaces

### PR DESCRIPTION
**Pull request by PawelLipski**
That was an awkward/poor idea on my part to keep abstract classes in the API... maybe that was super-correct wrt. equality contract, but also made it super-unwieldy wrt. extracting common implementation b/w implementations (see esp. `BaseGitMacheteBranch` now)

### CODE REVIEW

File: backendImpl/src/main/java/com/virtuslab/gitmachete/backend/impl/BaseGitMacheteBranch.java [hunk 1]
```diff
@@ -0,0 +1,60 @@
...
+import org.checkerframework.checker.nullness.qual.Nullable;
+
+import com.virtuslab.gitmachete.backend.api.IGitMacheteBranch;
+import com.virtuslab.gitmachete.backend.api.IGitMacheteCommit;
+import com.virtuslab.gitmachete.backend.api.IGitMacheteNonRootBranch;
+import com.virtuslab.gitmachete.backend.api.IGitMacheteRemoteBranch;
+import com.virtuslab.gitmachete.backend.api.SyncToRemoteStatus;
+
+@Getter
+public abstract class BaseGitMacheteBranch implements IGitMacheteBranch {
```
Comment (mkondratek) [inline backendImpl/src/main/java/com/virtuslab/gitmachete/backend/impl/BaseGitMacheteBranch.java:15]: Very good job! The many a field (:wink:) repeated in root and non root implementations was a real pain. :sob:

File: backendImpl/src/main/java/com/virtuslab/gitmachete/backend/impl/GitMacheteRepository.java [hunk 2]
```diff
@@ -19,7 +19,7 @@
   private static final IPrefixedLambdaLogger LOG = PrefixedLambdaLoggerFactory.getLogger("backend");
```
Comment (mkondratek) [inline backendImpl/src/main/java/com/virtuslab/gitmachete/backend/impl/GitMacheteRepository.java:19]: unused var?

File: gitCoreApi/src/main/java/com/virtuslab/gitcore/api/IGitCoreRepository.java [hunk 3]
```diff
@@ -22,5 +22,5 @@
 
   Path getMainDirectoryPath();
 
-  boolean isAncestor(BaseGitCoreCommit presumedAncestor, BaseGitCoreCommit presumedDescendant) throws GitCoreException;
+  boolean isAncestor(IGitCoreCommit presumedAncestor, IGitCoreCommit presumedDescendant) throws GitCoreException;
```
Comment (mkondratek) [inline gitCoreApi/src/main/java/com/virtuslab/gitcore/api/IGitCoreRepository.java:25]: while we are here, get rid of  `GitCoreException` from `getAllRemoteBranches()` 
(here and within `IGitCoreRepository`)

Comment (PawelLipski) [reply gitCoreApi/src/main/java/com/virtuslab/gitcore/api/IGitCoreRepository.java:25]: The problem is... `com.virtuslab.gitcore.impl.jgit.GitCoreRepository#getAllRemoteBranches` does throw `GitCoreException`, it's just "sneakified" by `Try.of(...).get()` (actually, pretty much equivalent to `@lombok.SneakyThrows`/`lombok.Lombok.sneakyThrow`)

Comment (PawelLipski) [reply gitCoreApi/src/main/java/com/virtuslab/gitcore/api/IGitCoreRepository.java:25]: I reworked it now with [`Try.traverse`](https://www.javadoc.io/doc/io.vavr/vavr/0.10.0/io/vavr/control/Try.html#traverse-java.lang.Iterable-java.util.function.Function-), it's more verbose but also clearer/more correct

**Comment by PawelLipski** [comment]
@mkondratek fixes pushed, also piggybacked a bunch of optimizations for local builds (4:30->2:30 for `./gradlew compileJava`)

File: CONTRIBUTING.md [hunk 4]
```diff
@@ -23,6 +23,9 @@ Most non-standard/project-specific conventions are enforced by:
 * [Checker Framework](https://checkerframework.org/manual/) for formal correctness, esp. wrt. null safety and UI thread handling
   (most config in [build.gradle](build.gradle), stubs in [config/checker/](config/checker))
 
+Note that certain Checker Framework's checkers (`Index`, `Optional`, `Regex`) are by default only run in the CI to speed up local compilation. <br/>
```
Comment (mkondratek) [inline CONTRIBUTING.md:26]: :hourglass:

File: gitCoreJGit/src/main/java/com/virtuslab/gitcore/impl/jgit/GitCoreRepository.java [hunk 5]
```diff
@@ -133,12 +134,19 @@ public IGitCoreLocalBranch getLocalBranch(String branchName) throws GitCoreExcep
...
+  }
+
+  private Try<List<IGitCoreRemoteBranch>> getRemoteBranchesTry(String remoteName) {
+    return Try.of(() -> getRemoteBranches(remoteName));
   }
 
   @Override
   public List<IGitCoreRemoteBranch> getAllRemoteBranches() throws GitCoreException {
-    return getRemotes().flatMap(remoteName -> Try.of(() -> getRemoteBranches(remoteName)).get());
+    return Try.traverse(getRemotes(), this::getRemoteBranchesTry)
```
Comment (mkondratek) [inline gitCoreJGit/src/main/java/com/virtuslab/gitcore/impl/jgit/GitCoreRepository.java:146]: :heart:

**Pull request merged**